core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/ContainerTorrentDownloadJob.kt

```diff
@@ -0,0 +1,218 @@
+                            }
+                        }
+                    }
+
+                    ustadTorrentManager.addDownloadListener(containerUid, torrentListener as TorrentDownloadListener)
+                    torrentDeferred.await()
+                    ustadTorrentManager.removeDownloadListener(containerUid)
+
+                }
+
+
+                val existingMd5s = existingFiles.mapNotNull { it.cefMd5 }.toSet()
+                val entriesThatGotDownloaded = fileNameList
+                        .filter { it !in existingMd5s || it.equals(MANIFEST_FILE_NAME) }
+
+                val newContainerEntryFiles = mutableListOf<ContainerEntryFile>()
+                entriesThatGotDownloaded.forEach {
+                    val entryFile = ContainerEntryFile().apply {
+                        cefMd5 = it
+                        val cefFile = File(containerFilesFolder, it)
+                        cefPath = cefFile.path
+                        ceTotalSize = cefFile.getUnCompressedSize()
+                        ceCompressedSize = cefFile.length()
+                        compression = if (cefFile.isGzipped()) COMPRESSION_GZIP else COMPRESSION_NONE
+                    }
+                    newContainerEntryFiles.add(entryFile)
+                }
+                db.containerEntryFileDao.insertList(newContainerEntryFiles)
+
+                val manifestFile = File(containerFilesFolder, MANIFEST_FILE_NAME)
+                if (!manifestFile.exists()) throw IllegalArgumentException("no manifest file found")
+
+                val manifestJson = manifestFile.readText()
+                val manifest: ContainerManifest = Json.decodeFromString(
+                        ContainerManifest.serializer(), manifestJson)
+                val newContainerEntry = mutableListOf<ContainerEntry>()
+                manifest.entryMap?.entries?.forEach { entry ->
+                    if (entry.key == MANIFEST_FILE_NAME) {
+                        return@forEach
+                    }
+                    entry.value.forEach { path ->
+                        val cefUid = db.containerEntryFileDao.findEntryByMd5Sum(entry.key.base64EncodedToHexString())?.cefUid
+                                ?: throw IllegalArgumentException("missed a file during download ${entry.key} with path $path")
+                        val containerEntry = ContainerEntry().apply {
+                            ceContainerUid = containerUid
+                            cePath = path
+                            ceCefUid = cefUid
+                        }
+                        newContainerEntry.add(containerEntry)
+                    }
+                }
+                db.containerEntryDao.insertListAsync(newContainerEntry)
+
+                contentJobItem.cjiItemProgress = contentJobItem.cjiItemTotal
+                progress.onProgress(contentJobItem)
+
+            } catch (c: CancellationException) {
+                throw c
```

Comment:
If the torrent is not downloaded, this must remove the torrent (but should not delete the torrent file itself)!